### DISCUSSION TITLE
Spreading awareness of PEP 752

### DISCUSSION THREAD

**Discussion by ofek** [mkdocs/mkdocs#3810] [Meta]
Hello team! I just wanted to bring to your attention a new PEP about package repository namespaces where I used your packages/community as a motivating example 🙂

We would love your feedback in the current discussion: https://discuss.python.org/t/pep-752-package-repository-namespaces/61227

**Comment by pawamoy** [comment]
Nice! Great job on the PEP @ofek, and thanks for sharing here 🙂

So, as said in the discussion, the only benefit to public namespaces (for example `mkdocs-*`), is a visual indicator that a package is "official" when it is published by an account that is part of the owners of the namespace (IIUC).

Let say MkDocs reserves the `mkdocs-*` namespace, and I'm part of the owners: would my own, *personal* MkDocs plugins get the visual indicator? Even though they are not at all endorsed in any way by the MkDocs org? Even the ones I published *before* I was added to the org / listed as owner?

I can see a few scenarios in which this would give a false sense of security.

Other than that, non-public namespaces make a lot of sense to me, in regard to orgs who want complete control over the prefix 🙂 I'm just not really the target audience for those though.

**Reply by ofek** [reply]
That's a very good, edge casey question! No the indicator would be tied to the authentication mechanism i.e. must be a key reserved specifically for the organization/project. I will update the proposal to make that explicit, thanks! Also, if you feel this would benefit (or not) the MkDocs community overall it would be nice to get your take on it in the discussion. I'm posting this announcement in a bunch of projects that might benefit.

**Reply by pawamoy** [reply]
Yes, happy to discuss it there too, just wanted to reply with a kinda MkDocs-specific answer here 🙂

OK, thanks for your answer, so to get the visual indicator you'd have to publish with a special account / an account with special auth (as in, explicitly listed API key or something). That addresses my concerns 👍

As to whether it would benefit MkDocs, right now I would say "no" because MkDocs maintains only one or two plugins (mkdocs-redirects, the minify plugin... Something else?). But maybe if we started welcoming projects in the org and endorsing them, then yes, this could be interesting 🙂

I'll share all this in the thread later!

**Reply by ofek** [reply]
Also please note part one of the [semantics](https://peps.python.org/pep-0752/#grant-semantics) i.e. not only would those two packages become special but so would the base name `mkdocs`.

**Reply by lovelydinosaur** [reply]
> As to whether it would benefit MkDocs I would say "no"

I'm perhaps somewhat with Tim here.

Notions of ownership (relational attachment). Irrepressibly organic. Permeability.

Do we get to namespace f the name "mkdocs" is cited in a PEP. Which we? How would feel about that?

(I'm being deliberately squiggly there, hopefully the general sense comes across. Managing how we make decisions of who-has-access-to-what in open source projects like this is a challenge for sure. My personal take would be that strictly org-namespaced packages would introduce more issues than it would resolve, at least in our case.)

**Reply by ofek** [reply]
Sorry, could you please elaborate? I didn't quite understand what you meant.

**Reply by lovelydinosaur** [reply]
I've elaborated a little, in edit. (Hope that helps?)

**Reply by ofek** [reply]
That helps a bit, thanks! Just so I have full context, did both of you read the entire proposal?

**Reply by lovelydinosaur** [reply]
Mine was a cursory skim. Proposal & comments. Not sure about Tim's.

Interesting to me as have been looking into to what extent we might be able to move away from requiring Python packaging to support plugin functionality.

**Reply by lovelydinosaur** [reply]
Reviewed in full. Exceptional thoughtful & thorough. Appreciate your diligence.

Wrt feedback, I'd echo this from Jakub...

> I guess the question is: do other companies consider this helpful

Socially something that's a little fuzzy to me is how our team could meaningfully vet / vouch for deployments and updates to the multitude of projects under our namespace.

May keep in touch with this conversation as it evolves.

Thanks.

**Reply by pawamoy** [reply]
> Just so I have full context, did both of you read the entire proposal?

I did before answering, yes :relaxed:

**Comment by squidfunk** [comment]
@ofek we'd be very interested in package namespaces, but also vouch for npm-style scoping, so it's clear that things belong to an organization ☺️ I also thought about cutting my work into smaller packages, and scoping would perfectly allow for that, e.g.:

- `@mkdocs-material/core`
- `@mkdocs-material/plugin-*` (blog, tags, ...)
- `@mkdocs-material/icons-*` (material, octicons, fontawesome, ...)

This would also allow to create things like presets, e.g. in our case for blog, blog + documentation, knowledge base, etc:

- `@mkdocs-material/preset-blog`
- `@mkdocs-material/preset-docs`

Retrofitting scopes should be simple. Reserving prefixes in an already pretty crowded space might be challenging 😅

**Reply by pawamoy** [reply]
The PEP states that NPM-style scopes would be too disruptive to the ecosystem :sweat_smile: (think about all the package installers that would now need to support scopes). It also says that existing packages would be unimpacted even if a matching prefix is reserved, so that would make reserving a prefix straightforward :slightly_smiling_face:

**Reply by squidfunk** [reply]
I'm aware that support needs to be added to installers (as was added to npm), but I think it should be possible to implement this as a backward compatible change, so I wouldn't see much of a problem here. As there are already million (?) of Python packages, reserving a mint prefix is almost impossible, so you'd always be in chartered territory. This means that although I might start with a package prefix identifier like "foobar-*"  to release my packages, as soon as somebody reserves this prefix, I wouldn't be able to publish new packages under it, right?

However, just my 2 cents 🤷‍♂️ Quite hard to find free names (or prefixes) on PyPI right now 😅

**Reply by pawamoy** [reply]
> This means that although I might start with a package prefix identifier like "foobar-*" to release my packages, as soon as somebody reserves this prefix, I wouldn't be able to publish new packages under it, right?

Right, unless the namespace is public, meaning anyone can continue publishing new projects under it :slightly_smiling_face: But yeah, valid concern :thinking:

**Reply by squidfunk** [reply]
Also, what qualifies a prefix for being reserved? So if `mkdocs` would be reserved right now, there are around 300 packages shipping third party plugins with `mkdocs` as a prefix. Thus, once `mkdocs-*` is reserved, nobody could publish new `mkdocs-*` packages. Thus, from a maintainer's point of view trying to build an ecosystem (like MkDocs successfully did), the best thing would actually be not to reserve the package prefix to foster collaboration and community driven plugins.

**Reply by pawamoy** [reply]
The PEP says that community orgs (which MkDocs would be unless it starts paying) SHALL always use public namespaces. I understand it as "community orgs will usually only be granted public namespaces, but we might make exceptions"?

In any case I think too that non-public namespaces for open source orgs and communities seem counter-productive.

Unless maybe when it's a new org and it clearly communicates a naming pattern from the start, like `foo-*` is reserved and third-party should use `fooplugin-*`.

**Reply by squidfunk** [reply]
To add to that, the fact that a namespace _must be granted_ will make things even more complicated. Just allowing `@my-namespace/` seems like the most sensible and open approach, allowing for paying members to publish private packages. Just like npm. I'm sorry if that's too narrow or unpythonic of a view, but the given proposal seems problematic to me.
